src/Rings/groebner.jl

```diff
@@ -205,58 +205,57 @@ end
 
 # leading ideal #######################################################
 @doc Markdown.doc"""
-    leading_ideal(g::Vector{T}, args...) where { T <: MPolyElem }
+    leading_ideal(g::Vector{T}; ordering::MonomialOrdering) where { T <: MPolyElem }
 
-Return the ideal generated by the leading monomials of the given polynomials.
-If not otherwise given as a further argument this is done w.r.t. the
-degree reverse lexicographical monomial ordering.
+Return the ideal generated by the leading monomials of the given polynomials
+w.r.t. the given monomial ordering.
 
 # Examples
 ```jldoctest
 julia> R, (x, y) = PolynomialRing(QQ, ["x", "y"])
 (Multivariate Polynomial Ring in x, y over Rational Field, fmpq_mpoly[x, y])
 
-julia> L = leading_ideal([x*y^2-3*x, x^3-14*y^5])
+julia> L = leading_ideal([x*y^2-3*x, x^3-14*y^5], ordering=degrevlex(gens(R)))
+ideal(x*y^2, y^5)
+
+julia> L = leading_ideal([x*y^2-3*x, x^3-14*y^5], ordering=lex(gens(R)))
 ideal(x*y^2, x^3)
 ```
 """
-function leading_ideal(g::Vector{T}, args...) where { T <: MPolyElem }
-  return ideal([ leading_monomial(f, args...) for f in g ])
+function leading_ideal(g::Vector{T}; ordering::MonomialOrdering) where { T <: MPolyElem }
+    return ideal(parent(g[1]), [first(monomials(f, ordering)) for f in g])
 end
 
 @doc Markdown.doc"""
-    leading_ideal(g::Vector{Any}, args...)
+    leading_ideal(g::Vector{Any}; ordering::MonomialOrdering)
 
 Given an array of fitting type this function returns the ideal
-generated by the leading monomials of the given array elements.
-If not otherwise given as a further argument this is done w.r.t.
-the degree reverse lexicographical monomial ordering.
+generated by the leading monomials of the given array elements
+w.r.t. the given monomial ordering.
 """
-function leading_ideal(g::Vector{Any}, args...)
-  return leading_ideal(typeof(g[1])[ f for f in g ], args...)
+function leading_ideal(g::Vector{Any}; ordering::MonomialOrdering)
```

Comment:
Note that keyword arguments do *not* participate in method dispatch.

So in effect, the above method turns `ordering` into a required keyword argument. Is that the intention?

I am also confused as to why a `Vector{Any}` is expected here. I would have expected a Vector of polynomials. What "array elements" are to be used here then?

Reply:
Argh, of course the PR description says "enforce a given monomial ordering" so yes, you do this intentionally.

So why is it a keyword argument?

Reply:
We (@wdecker) want to have the orderings as keywords for the whole gb related stuff. It is also a keyword for `groebner_basis`and should be one for `normal_form`, etc.

Reply:
And I removed the `Vector{Any}` instantiation of `leading_ideal`, it was deprecated anyway.